src/test/ui/resolve/issue-103202.rs

```diff
@@ -0,0 +1,7 @@
+struct S {}
+
+impl S {
+    fn f(self: &S::x) {} //~ ERROR ambiguous associated type
```

Comment:
Hmm.. that error seems incorrect. Maybe we should bubble out a new `ResolutionError` variant from `find_lifetime_for_self` by changing the `lifetime` field of the visitor to a `Result` that, once set to `Err` simply doesn't do anything else anymore.

Reply:
The error does not come from `find_lifetime_for`self`, it looks like the regular error because `S` is known to be a type and we expect a type.

Reply:
Before this regression, rustc caused the following error:

```
error[E0223]: ambiguous associated type
 --> src/main.rs:4:17
  |
4 |     fn f(self: &S::x) {} //~ ERROR ambiguous associated type
  |                 ^^^^ help: use fully-qualified syntax: `<S as Trait>::x`
```